sklearn/utils/_random.pyx

```diff
@@ -303,6 +308,29 @@ cpdef sample_without_replacement(cnp.int_t n_population,
                          % (all_methods, method))
 
 
+def sample_without_replacement(
+        object n_population, object n_samples, method="auto", random_state=None):
+    cdef:
+        cnp.intp_t n_pop_intp, n_samples_intp
+        long n_pop_long, n_samples_long
+
+    # Note: the legacy rng will use `long` internally in either case
+    # If NumPy 2.0 is guaranteed at compile time this could also be written
+    # as `cnp.NPY_DEFAULT_INT == cnp.NPY_INTP`.
+    if np.int_ is np.intp:
```

Comment:
Actually, the CI hacks non-withstanding.  `np.int_ is np.intp` on most platforms, and the second branch is for *old* versions of NumPy on windows.  So CI is already covering both branches here.

In other words:  The trial may be interesting to see if there are more issues, but as soon as anyone reviews this and thinks it's fine, I would propose to just merge it even if the trials get stuck.

Reply:
May you add a comment about Windows + old NumPy versions in the second branch and also the comment about `np.int_ is np.intp` on most platforms.